### ISSUE TITLE
keras_CNN_mnist.ipynb

### ISSUE THREAD

**Issue by OoallenoO123123123** [OoallenoO123123123/AI-#3]
# 匯入必要模組
from tensorflow import keras
from keras.models import Sequential
from keras.layers import Dense, Conv2D, MaxPooling2D, Flatten

# ========================
# 建立 CNN 模型
# ========================
model=Sequential()

# 第一層：卷積層 (Conv2D)
# filters=16：建立 16 個卷積核 (特徵偵測器)
# kernel_size=(5,5)：每個卷積核大小為 5x5
# input_shape=(28,28,1)：輸入影像尺寸 (灰階單通道)
# activation='relu'：啟用 ReLU 激活函數
model.add(Conv2D(filters=16, kernel_size=(5, 5), input_shape=(28, 28, 1), activation='relu'))

# 第二層：最大池化層 (MaxPooling2D)
# pool_size=(2,2)：將特徵圖縮小一半以減少參數數量
model.add(MaxPooling2D(pool_size=(2, 2)))
#model.add(Conv2D(filters=36, kernel_size=(5, 5), activation='relu'))
#model.add(MaxPooling2D(pool_size=(2, 2)))

# Flatten：將多維特徵圖展平成一維輸入向量
model.add(Flatten())

# 輸出層：10 個神經元代表數字 0~9
# activation='softmax'：輸出為分類機率
model.add(Dense(10, activation='softmax'))

# 顯示模型架構摘要
model.summary()

# 編譯模型
# optimizer='rmsprop'：使用 RMSprop 最佳化器
# loss='categorical_crossentropy'：多類別分類的損失函數
# metrics=['accuracy']：以準確率衡量模型效能
model.compile(optimizer='rmsprop', loss='categorical_crossentropy', metrics=['accuracy'])

Model: "sequential"
_________________________________________________________________
 Layer (type)                Output Shape              Param #   
=================================================================
 conv2d (Conv2D)             (None, 24, 24, 16)        416       
                                                                 
 max_pooling2d (MaxPooling2D  (None, 12, 12, 16)       0         
 )                                                               
                                                                 
 flatten (Flatten)           (None, 2304)              0         
                                                                 
 dense (Dense)               (None, 10)                23050     
                                                                 
=================================================================
Total params: 23,466
Trainable params: 23,466
Non-trainable params: 0

# ========================
# 讀取與前處理資料
# ========================
mnist=keras.datasets.mnist
(train_data, train_label), (test_data, test_label) = mnist.load_data()

# 保存原始標籤，用於混淆矩陣
test_label_original=test_label #for crosstab

# one-hot 編碼：將標籤轉為10維向量（如「3」→ [0,0,0,1,0,0,0,0,0,0]）
train_label = keras.utils.to_categorical(train_label)
test_label = keras.utils.to_categorical(test_label)

# ========================
# 模型訓練
# ========================
# epochs=10：訓練10次
# batch_size=512：每次用512張影像更新權重
train_history=model.fit(train_data,train_label,epochs=10,batch_size=512)

Epoch 1/10
118/118 [==============================] - 5s 15ms/step - loss: 3.9564 - accuracy: 0.8308
Epoch 2/10
118/118 [==============================] - 2s 14ms/step - loss: 0.2603 - accuracy: 0.9589
Epoch 3/10
118/118 [==============================] - 2s 14ms/step - loss: 0.1189 - accuracy: 0.9740
Epoch 4/10
118/118 [==============================] - 2s 15ms/step - loss: 0.0754 - accuracy: 0.9816
Epoch 5/10
118/118 [==============================] - 2s 15ms/step - loss: 0.0511 - accuracy: 0.9863
Epoch 6/10
118/118 [==============================] - 2s 15ms/step - loss: 0.0386 - accuracy: 0.9893
Epoch 7/10
118/118 [==============================] - 2s 13ms/step - loss: 0.0295 - accuracy: 0.9914
Epoch 8/10
118/118 [==============================] - 2s 13ms/step - loss: 0.0237 - accuracy: 0.9928
Epoch 9/10
118/118 [==============================] - 1s 10ms/step - loss: 0.0199 - accuracy: 0.9940
Epoch 10/10
118/118 [==============================] - 1s 10ms/step - loss: 0.0171 - accuracy: 0.9948

# ========================
# 模型評估
# ========================
score=model.evaluate(test_data, test_label)
print(score[1])  # 輸出準確率（約 0.977 ≈ 97.7%）

313/313 [==============================] - 1s 4ms/step - loss: 0.1323 - accuracy: 0.9774
0.977400004863739

#繪製正確性與損失立歷史圖
import matplotlib.pyplot as plt
import numpy as np

def show_train_history(train_history, train, validation):
    plt.plot(train_history.history[train])
    plt.plot(train_history.history[validation])
    plt.title('Train History')
    plt.ylabel('Train')
    plt.xlabel('Epoch')
    plt.legend(['train', 'validation'], loc='center right')
    plt.show()

# 繪出訓練準確率與損失變化
show_train_history(train_history, 'accuracy', 'accuracy')
show_train_history(train_history, 'loss', 'loss')

# Confusion Matrix混淆矩陣
import pandas as pd

# ========================
# 混淆矩陣 (Confusion Matrix)
# ========================
# 預測測試資料
prediction = model.predict(test_data)

# 取出機率最高的類別作為預測標籤
prediction_label=np.argmax(prediction,axis=1)
print(test_label.shape)

# 列出實際標籤 vs 預測標籤的交叉表
pd.crosstab(test_label_original, prediction_label, rownames=['label'], colnames=['predict'])
